### PR TITLE
L1 emulator for LLP displaced jet trigger (L1 6:1 LUT, L2 jet algo), L1TNtuples addition of jet hwQual

### PR DESCRIPTION
**Pull request by gk199**
#### PR description:

This PR implements the L1 emulator of the LLP displaced jet trigger algorithm. Calo L1 takes the 6 fine grain bits from the HCAL uHTR (PRs [#34600](https://github.com/cms-sw/cmssw/pull/34600) and [#35599](https://github.com/cms-sw/cmssw/pull/35599)) and uses a 6:1 LUT to set a single bit of tower hwQual based on the logical OR of the timing and depth information. The Calo L2 jet algorithm flags a jet as delayed if there are >= 2 flagged towers within the 9x9 jet region. This is saved in the LSB of jet.hwQual, 1 if jet is flagged as containing an LLP, 0 otherwise. L1 and L2 firmware updates have been completed. 

The uGT emulator follows, and is implemented in recent PR [#36868](https://github.com/cms-sw/cmssw/pull/36858). L1 Menu Tools is in the process of being updated to include the displaced jet bit, in PR [#74](https://github.com/cms-l1-dpg/L1MenuTools/pull/74).

The L1 seed JIRA ticket, including slides detailing the testing of the L1 emulator and links to relevant presentations, is [here](https://its.cern.ch/jira/browse/CMSLITDPG-963). Slides 2-10 focus on the L1 emulator implementation and validation tests. 

The emulator implementation has been presented at [L1 Trigger CMS Week](https://indico.cern.ch/event/1072305/contributions/4513777/attachments/2308783/3928427/HCAL_L1TCMSweek_14Sept.pdf) and to the [HCAL Trigger](https://indico.cern.ch/event/1069597/contributions/4497838/attachments/2299267/3914153/HCALtrigger_27August.pdf) group.

Additionally, the L1TNtuples are updated to store the jet hwQual information (in `L1AnalysisL1Upgrade.cc` and `L1AnalysisL1UpgradeDataFormat.h`). Tested by myself and @elfontan.

A single comment in `HcalFinegrainBit.cc` is updated to reflect the correct assignment of the 6 fine grain bits from the uHTR. 

#### PR validation:

Passed runTheMatrix.py tests. Tested in CMSSW_12_3_X_2022-02-08-1100.

### CODE REVIEW

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by cecilecaillol** [comment]
@perrotta @dpiparo @qliphy Can this be merged or would you like us to address some issues?

**Comment by qliphy** [comment]
@cecilecaillol  We will cut pre5 soon, and now are waiting for CMSSW_12_3_X_2022-02-16-1100 IB to be done.  We will include this afterwards if this PR is not urgent for pre5.   
BTW:  the PR tests differences are as expected, right?

**Comment by cecilecaillol** [comment]
@qliphy We would like to have this PR in pre5 if still possible because we would like to synthetise the new L1 menu for tests with the uGT test crate and we need to check the agreement between firmware and emulator, and it would be the best to prepare the test vector for the validation with the latest official emulator. Adding @elfontan if more information is needed

File: L1Trigger/L1TCaloLayer1/plugins/L1TCaloLayer1.cc [hunk 1]
```diff
@@ -209,16 +209,23 @@ void L1TCaloLayer1::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
-          if (fg)
-            featureBits |= 0b01;
-          // fg2 should only be set for HF
-          if (absCaloEta > 29 && fg2)
-            featureBits |= 0b10;
+          if (absCaloEta > 29) {
+            if (fg)
+              featureBits |= 0b01;
+            // fg2 should only be set for HF
+            if (absCaloEta > 29 && fg2)
```
Comment (perrotta) [inline L1Trigger/L1TCaloLayer1/plugins/L1TCaloLayer1.cc:224]:
```suggestion
            if (fg2)
```

File: L1Trigger/L1TCaloLayer1/plugins/L1TCaloLayer1.cc [hunk 2]
```diff
@@ -209,16 +209,23 @@ void L1TCaloLayer1::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
...
+              featureBits |= 0b10;
+          }
+          if (absCaloEta <= 29)
```
Comment (perrotta) [inline L1Trigger/L1TCaloLayer1/plugins/L1TCaloLayer1.cc:227]:
```suggestion
          else
```

**Comment by elfontan** [comment]
@qliphy, I would like to also add that the other PR for displaced jets related to the development at the uGT level has been already merged ([PR#36858](https://github.com/cms-sw/cmssw/pull/36858)), so it would be good to make sure that we have the full integration of the emulator for the displaced jets (including Calo emulator) in the new release.

**Comment by perrotta** [comment]
> @qliphy We would like to have this PR in pre5 if still possible because we would like to synthetise the new L1 menu for tests with the uGT test crate and we need to check the agreement between firmware and emulator, and it would be the best to prepare the test vector for the validation with the latest official emulator. Adding @elfontan if more information is needed

@cecilecaillol you are certainly aware that we have every Tuesday at 5pm an Offline Release Planning meeting, where the integrations are discussed, and priorities and urgencies are raised, if any. In particular yesterday we had the final ORP meeting before 12_3_0_pre5: whoever had urgency to have code integrated in pre5 signalled it at the meeting, and we did our best to integrate everything in time. I am afraid we did not get any PR pointed out by L1T: please either organize yourself to participate to the meeting, send a delegate, or send a message in advance with your requests.

As of now, the build for 12_3_0_pre5 is ready to start, and we don't plan to block it, unless you insist that having this update in pre5 instead of pre6 is really vital for your ongoing developments.

**Review by perrotta — COMMENTED** [review]
Please comment about the differences observed in the L1T DQM outputs: I imagine they are as expected, but it is not clear from the PR description and the discussion in this github thread

**Comment by cecilecaillol** [comment]
@perrotta We are sorry for not mentioning it at the meeting yesterday (which we attended), we had an internal miscommunication and didnt realize it was urgent. However, we would really appreciate a lot if this could be included in pre5 because it will be a big problem and a lot of work more for our menu subgroup otherwise. Is this still possible? Sorry!

**Comment by cecilecaillol** [comment]
+l1

**Comment by perrotta** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by perrotta** [comment]
@cecilecaillol build for pre5 has started, pre6 will be in only three weeks from now. I suppose you can still keep developing by `git merge topic` this PR on top of pre5 in your working area during these three weeks.

In the meanwhile, if anyone can answer about the correctness of the L1T DQM comparison, it could simplify accepting this PR at once as soon as the bot tests have finished. The tiny differences observed are all in some jet-qual related plot, and therefore I would say that at least this PR touches where it should. But an expert opinion on it is always appreciated.

**Comment by cecilecaillol** [comment]
@perrotta ok, thanks! @gk199 will comment about the DQM comparison

**Comment by gk199** [comment]
L1T DQM histogram comparison comments:
- Changes are seen in the jet QUAL bit plot (CenJetsQual), and are expected. The `Stage2Layer2JetAlgorithmFirmwareImp1.cc` sets the least significant bit of jet hwQual (to 1) if the jet is found to be delayed based on the HCAL feature bits. This bit is set infrequently, as seen in the central jet quality histogram.
- In the DQM histograms, I also see the HCAL Feature Bit occupancy histograms at Layer 1. This is also expected since the feature bit is now being set based a 6:1 LUT from the 6 fine grain bits from the HCAL uHTR (update to Layer-1 logic, as previously the feature bit just relied on the first uHTR fine grain bit).

**Comment by perrotta** [comment]
+1

**Comment by gk199** [comment]
> We're getting ASAN heap-buffer-overflow errors, likely related to this PR, see #37012

I added some suggestions above in `CaloLayer1Packer.cc`, please let me know if there are any other thoughts to resolve this (or input from someone from L1, @cecilecaillol?). Is there a way to test the ASAN build locally?

**Comment by makortel** [comment]
> Is there a way to test the ASAN build locally, with a command to run to reproduce the error?

Yes, you can check the latest ASAN build from the [IB dashboard](https://cmssdt.cern.ch/SDT/html/cmssdt-ib) and run a failing workflow there, e.g.
```
cmsrel CMSSW_12_3_ASAN_X_2022-02-28-1100
cd CMSSW_12_3_ASAN_X_2022-02-28-1100/src
cmsenv
runTheMatrix.py -l 23234.0
```

**Comment by cecilecaillol** [comment]
@gk199 Could you check if your suggested fixes above remove the errors?

**Comment by gk199** [comment]
> @gk199 Could you check if your suggested fixes above remove the errors?

I tested my suggested fix, and still have the heap-buffer-overflow issue. I also tested a number of other modifications, such as doing the 6:1 logic in CaloLayer1Packer.cc prior to passing through setFB instead of doing the 6:1 logic in UCTCTP7RawData.h. 

Any other suggestions or guidance will be much appreciated as I work to resolve this.

**Comment by dan131riley** [comment]
This loop:
https://github.com/cms-sw/cmssw/blob/62d26a25201604dae1a25c3ccf434613cda1cc18/EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Packer.cc#L111-L113
needs to be inside this conditional:
https://github.com/cms-sw/cmssw/blob/62d26a25201604dae1a25c3ccf434613cda1cc18/EventFilter/L1TRawToDigi/plugins/implementations_stage2/CaloLayer1Packer.cc#L115-L118
As is, when `id` isn't found it's indexing off the end of the vector.

**Comment by gk199** [comment]
Many thanks @dan131riley for the fix and the explanation! I confirm this resolves the step2 heap-buffer-overflow from `CaloLayer1Packer::makeHCalTPGs` when running `runTheMatrix.py -l 39834.0` in CMSSW_12_4_ASAN_X_2022-03-14-1100. 

However, I see an error in step3, which seems unrelated to any code in this PR, as it is part of `HGVHistoProducerAlgo`. 
```==6890==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60d008945b0c at pc 0x7fd8a67a0b48 bp 0x7fff748f3790 sp 0x7fff748f3788
READ of size 4 at 0x60d008945b0c thread T0
    #0 0x7fd8a67a0b47 in HGVHistoProducerAlgo::tracksters_to_SimTracksters(HGVHistoProducerAlgoHistograms const&, int, std::vector<ticl::Trackster, std::allocator<ticl::Trackster> > const&, std::vector<reco::CaloCluster, std::allocator<reco::CaloCluster> > const&, std::vector<ticl::Trackster, std::allocator<ticl::Trackster> > const&, HGVHistoProducerAlgo::validationType, std::vector<ticl::Trackster, std::allocator<ticl::Trackster> > const&, std::map<unsigned int, std::vector<unsigned int, std::allocator<unsigned int> >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::vector<unsigned int, std::allocator<unsigned int> > > > > const&, std::vector<SimCluster, std::allocator<SimCluster> > const&, edm::ProductID const&, std::vector<CaloParticle, std::allocator<CaloParticle> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::unordered_map<DetId, HGCRecHit const*, std::hash<DetId>, std::equal_to<DetId>, std::allocator<std::pair<DetId const, HGCRecHit const*> > > const&, unsigned int) const (/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_4_ASAN_X_2022-03-14-1100/biglib/slc7_amd64_gcc11/pluginSimulation.so+0x17eeb47)
    #1 0x7fd8a67a713b in HGVHistoProducerAlgo::fill_trackster_histos(HGVHistoProducerAlgoHistograms const&, int, std::vector<ticl::Trackster, std::allocator<ticl::Trackster> > const&, std::vector<reco::CaloCluster, std::allocator<reco::CaloCluster> > const&, std::vector<ticl::Trackster, std::allocator<ticl::Trackster> > const&, std::vector<ticl::Trackster, std::allocator<ticl::Trackster> > const&, std::map<unsigned int, std::vector<unsigned int, std::allocator<unsigned int> >, std::less<unsigned int>, std::allocator<std::pair<unsigned int const, std::vector<unsigned int, std::allocator<unsigned int> > > > > const&, std::vector<SimCluster, std::allocator<SimCluster> > const&, edm::ProductID const&, std::vector<CaloParticle, std::allocator<CaloParticle> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::vector<unsigned long, std::allocator<unsigned long> > const&, std::unordered_map<DetId, HGCRecHit const*, std::hash<DetId>, std::equal_to<DetId>, std::allocator<std::pair<DetId const, HGCRecHit const*> > > const&, unsigned int) const (/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc11/cms/cmssw/CMSSW_12_4_ASAN_X_2022-03-14-1100/biglib/slc7_amd64_gcc11/pluginSimulation.so+0x17f513b)
```

As this is an unrelated file to this L1 emulator PR, may I now open a new PR to fix the heap-buffer-overflow with the changes in `CaloLayer1Packer::makeHCalTPGs`?

**Comment by dan131riley** [comment]
@gk199 please make a new PR with the step2 fix.  The fix is clearly correct and needed, the step3 error can be dealt with separately.

**Pull request merged**